### PR TITLE
python310Packages.pylint: 2.14.5 -> 2.15.5

### PR DESCRIPTION
**Pull request by tjni**
###### Description of changes

https://github.com/PyCQA/pylint/releases/tag/v2.15.5
https://github.com/PyCQA/pylint/releases/tag/v2.15.4
https://github.com/PyCQA/pylint/releases/tag/v2.15.3
https://github.com/PyCQA/pylint/releases/tag/v2.15.2
https://github.com/PyCQA/pylint/releases/tag/v2.15.1
https://github.com/PyCQA/pylint/releases/tag/v2.15.0

and

https://github.com/PyCQA/astroid/releases/tag/v2.12.12
https://github.com/PyCQA/astroid/releases/tag/v2.12.11
https://github.com/PyCQA/astroid/releases/tag/v2.12.10
https://github.com/PyCQA/astroid/releases/tag/v2.12.9
https://github.com/PyCQA/astroid/releases/tag/v2.12.8
https://github.com/PyCQA/astroid/releases/tag/v2.12.7
https://github.com/PyCQA/astroid/releases/tag/v2.12.6
https://github.com/PyCQA/astroid/releases/tag/v2.12.5
https://github.com/PyCQA/astroid/releases/tag/v2.12.4
https://github.com/PyCQA/astroid/releases/tag/v2.12.3
https://github.com/PyCQA/astroid/releases/tag/v2.12.2
https://github.com/PyCQA/astroid/releases/tag/v2.12.1
https://github.com/PyCQA/astroid/releases/tag/v2.12.0

Additional notes regarding the Z3/deal-solver change:

Had a conversation on Matrix that the `propagatedInputs` on the Z3 derivation are not automatically carried along even after wrapping `z3.python` within `toPythonModule`. For now, I've updated the derivations to pass them through and use them explicitly, but if this is ever changed to work automatically, this can be cleaned up.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by tjni** [comment]
@ofborg build python39Packages.pylint python310Packages.pylint

**Comment by tjni** [comment]
The astroid change is necessary for Python 3.11 and is needed by https://github.com/NixOS/nixpkgs/issues/198533.

File: pkgs/applications/science/logic/z3/default.nix [hunk 1]
```diff
@@ -35,6 +35,9 @@ let common = { version, sha256, patches ? [ ] }:
       ++ optionals ocamlBindings [ ocaml findlib zarith ]
     ;
     propagatedBuildInputs = [ python.pkgs.setuptools ];
+    passthru.optional-dependencies = {
+      python = [ python.pkgs.setuptools ];
+    };
```
Comment (SuperSandro2000) [inline pkgs/applications/science/logic/z3/default.nix:40]: passthru.optional-dependencies should only be used for extras that can be installed via ``pip install abc[extra]``.

File: pkgs/development/python-modules/astroid/default.nix [hunk 2]
```diff
@@ -45,11 +41,7 @@ buildPythonPackage rec {
 
   checkInputs = [
     pytestCheckHook
-  ];
-
-  disabledTests = [
-    # AssertionError: Lists differ: ['ABC[16 chars]yBase', 'Final', 'Generic', 'MyProtocol', 'Protocol', 'object'] != ['ABC[16 chars]yBase', 'Final', 'Generic', 'MyProtocol', 'object']
-    "test_mro_typing_extensions"
+    typing-extensions # see https://github.com/PyCQA/astroid/issues/1585
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/astroid/default.nix:44]:
```suggestion
    typing-extensions
```
already included in a commit https://github.com/PyCQA/astroid/commit/4e9225e7191fcb2a92eea35ce05e17d0660a8c6b

File: pkgs/applications/science/logic/z3/default.nix (hunk 1 above)
Comment (tjni) [reply pkgs/applications/science/logic/z3/default.nix]: Good to know. I switched to use `z3.requiredPythonModules`, set in `toPythonModule` to the `propagatedBuildInputs` of the Z3 derivation.

File: pkgs/development/python-modules/astroid/default.nix (hunk 2 above)
Comment (tjni) [reply pkgs/development/python-modules/astroid/default.nix]: Fixed.

**Comment by SuperSandro2000** [comment]
If there are any packages which fail to build after merging this because of new pylint checks, we should remove type checking from those packages.

**Pull request merged**